src/hotspot/share/opto/macro.cpp

```diff
@@ -1810,74 +1798,60 @@ void PhaseMacroExpand::jfr_sample_fast_object_allocation(
-      alloc_sample_enabled_mem = alloc_sample_hit_region_phi_mem;
-    }
-    alloc_sample_enabled_region->init_req(enabled_idx, alloc_sample_enabled_ctrl);
-    alloc_sample_enabled_region_phi_mem->init_req(enabled_idx, alloc_sample_enabled_mem);
-  }
-
-  {
-    alloc_sample_enabled_region->init_req(disabled_idx, alloc_sample_disabled_ctrl);
-    alloc_sample_enabled_region_phi_mem->init_req(disabled_idx, alloc_sample_disabled_mem);
-  }
+  enum { disabled_idx = 1, hit_idx = 2, miss_idx = 3 };
+  Node* result_region = transform_later(new RegionNode(4));
+  Node* result_phi_mem = transform_later(new PhiNode(result_region, Type::MEMORY, TypeRawPtr::BOTTOM));
+
+  const int alloc_count_offset = in_bytes(TRACE_THREAD_ALLOC_COUNT_OFFSET);
+  Node* alloc_count = make_load(alloc_sample_enabled_ctrl, alloc_sample_enabled_mem, tls, alloc_count_offset, TypeLong::LONG, T_LONG);
+  Node* alloc_count_new = transform_later(new AddLNode(alloc_count, longcon(1)));
+  alloc_sample_enabled_mem = make_store(alloc_sample_enabled_ctrl, alloc_sample_enabled_mem, tls, alloc_count_offset, alloc_count_new, T_LONG);
+  const int alloc_count_until_sample_offset = in_bytes(TRACE_THREAD_ALLOC_COUNT_UNTIL_SAMPLE_OFFSET);
+  Node* alloc_count_until_sample = make_load(alloc_sample_enabled_ctrl, alloc_sample_enabled_mem, tls, alloc_count_until_sample_offset, TypeLong::LONG, T_LONG);
+  Node* alloc_count_until_sample_cmp = transform_later(new CmpLNode(alloc_count_until_sample, alloc_count_new));
+  Node* alloc_sample_hit_bool = transform_later(new BoolNode(alloc_count_until_sample_cmp, BoolTest::eq));
+  IfNode* alloc_sample_hit_if = (IfNode*)transform_later(new IfNode(alloc_sample_enabled_ctrl, alloc_sample_hit_bool, PROB_MIN, COUNT_UNKNOWN));
+  Node* alloc_sample_hit_ctrl = transform_later(new IfTrueNode(alloc_sample_hit_if));
+  Node* alloc_sample_hit_mem = alloc_sample_enabled_mem;
+  Node* alloc_sample_miss_ctrl = transform_later(new IfFalseNode(alloc_sample_hit_if));
+  Node* alloc_sample_miss_mem = alloc_sample_enabled_mem;
+
+  // if sample_hit then
+  address addr = OptoRuntime::jfr_fast_object_alloc_Java();
+  CallStaticJavaNode *call = new CallStaticJavaNode(OptoRuntime::jfr_fast_object_alloc_Type(),
+                                                        addr,
+                                                        OptoRuntime::stub_name(addr),
+                                                        alloc->jvms()->bci(),
+                                                        TypeRawPtr::BOTTOM);
+
+  call->init_req(TypeFunc::Control, alloc_sample_hit_ctrl);
+  call->init_req(TypeFunc::I_O    , alloc->in(TypeFunc::I_O));
+  call->init_req(TypeFunc::Memory , alloc_sample_hit_mem);
+  call->init_req(TypeFunc::ReturnAdr, alloc->in(TypeFunc::ReturnAdr));
+  call->init_req(TypeFunc::FramePtr, alloc->in(TypeFunc::FramePtr));
+
+  call->init_req(TypeFunc::Parms+0, fast_oop);
+
+  copy_call_debug_info(alloc, call);
+  call->set_cnt(PROB_FAIR);
```

Comment:
PROB_FAIR means 50% percent. Is it correct value?

Reply:
I think this value will affect the layout of the code, but because we can't evaluate the interval set by the user, so I set a conservative value here.

Reply:
After offline discussion with @kuaiwei, we decided not to set `cnt`, the default value is `COUNT_UNKNOWN` which is more appropriate here.